File: src/main/kotlin/inflearn/java/_2_3_가위바위보.java
package inflearn.java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.security.InvalidParameterException;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class _2_3_가위바위보 {

    private static String match(int a, int b) throws InvalidParameterException{
        // 비기는 경우
        if(a == b) return "D";
        // A가 이기는 경우
        if(a == 1 && b == 3) return "A";
        if(a == 2 && b == 1) return "A";
        if(a == 3 && b == 2) return "A";
        // B가 이기는 경우
        // return "B"; //이렇게 끝나도 충분하지만
        // 만일 a == 2 && b == 1 의 경우를 생각하지 못한 상태여서 작성하지 않았다면
        // `A가 이기는 경우`임에도 "B"를 반환하는 상황이 생긴다.
        // 그래서 실제로는 `else`를 함부로 사용하면 안 되는 것 같다.
        if(b == 1 && a == 3) return "B";
        if(b == 2 && a == 1) return "B";
        if(b == 3 && a == 2) return "B";

        throw new InvalidParameterException();
    }

    public static List<String> solution(int[] inputA, int[] inputB) {
        List<String> answer = new ArrayList<String>();

        int len = inputA.length;

        for (int i = 0; i < len; i++) {
            answer.add(match(inputA[i], inputB[i]));
        }

        return answer;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] inputA = new int[n], inputB = new int[n];
        StringTokenizer stA = new StringTokenizer(br.readLine());
        StringTokenizer stB = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            inputA[i] = Integer.parseInt(stA.nextToken());
            inputB[i] = Integer.parseInt(stB.nextToken());
        }

        List<String> ans = solution(inputA, inputB);
        for(String result : ans) {
            System.out.println(result);
        }
    }
}

File: src/main/kotlin/inflearn/2_3 가위바위보.kt
package inflearn

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.security.InvalidParameterException
import java.util.*

object `2_3 가위바위보` {
    @Throws(InvalidParameterException::class)
    private fun match(a: Int, b: Int): String {
        if (a == 1) {
            if (b == 1) return "D" else if (b == 2) return "B" else if (b == 3) return "A"
        } else if (a == 2) {
            if (b == 1) return "A" else if (b == 2) return "D" else if (b == 3) return "B"
        } else if (a == 3) {
            if (b == 1) return "B" else if (b == 2) return "A" else if (b == 3) return "D"
        }
        throw InvalidParameterException()
    }

    fun solution(inputA: IntArray, inputB: IntArray): List<String> {
        val answer: MutableList<String> = ArrayList()
        val len = inputA.size
        for (i in 0 until len) {
            answer.add(match(inputA[i], inputB[i]))
        }
        return answer
    }

    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val br = BufferedReader(InputStreamReader(System.`in`))
        val n = br.readLine().toInt()
        val inputA = IntArray(n)
        val inputB = IntArray(n)
        val stA = StringTokenizer(br.readLine())
        val stB = StringTokenizer(br.readLine())
        for (i in 0 until n) {
            inputA[i] = stA.nextToken().toInt()
            inputB[i] = stB.nextToken().toInt()
        }
        val ans = solution(inputA, inputB)
        for (result in ans) {
            println(result)
        }
    }
}